### PR TITLE
Cache element_dofs property in basis

### PR DESCRIPTION
**Pull request by topikuu**
This improves performance in almost all cases.

Signed-off-by: Topi Kuutela <topi.kuutela@aalto.fi>

### CODE REVIEW

**Comment by topikuu** [comment]
element_dofs can't be turned into a functools.cached_property as the basis.tind is modified at some points, breaking its evaluation.

Where should the HashableNdArray be put? The roof folder is probably not the right location.

**Comment by kinnala** [comment]
Where is `self.tind` modified?

**Comment by kinnala** [comment]
Alright, it has something to do with the initialization order.

**Comment by kinnala** [comment]
I think it could be possible to refactor the constructors so that `self.tind` is basically always a constant whenever `self.element_dofs` gets called.

**Comment by kinnala** [comment]
New:
```python
In [2]: m = MeshTet().refined(5)
In [6]: basis = InteriorBasis(m, ElementTetP2())                                
In [7]: %timeit laplace.assemble(basis)                                         
2.83 s ± 25.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Old:
```python
In [7]: %timeit laplace.assemble(basis)                                         
3.35 s ± 36.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

File: skfem/assembly/basis/basis.py [hunk 1]
```diff
@@ -97,8 +99,15 @@ def N(self):
     @property
     def element_dofs(self):
         if self.tind is None:
+            return self.get_element_dofs(None)
+        else:
```
Comment (kinnala) [inline skfem/assembly/basis/basis.py:103]: `else` is unnecessary?

File: skfem/assembly/basis/basis.py [hunk 2]
```diff
@@ -97,8 +99,15 @@ def N(self):
...
+            return self.get_element_dofs(HashableNdArray(self.tind))
+
+    @lru_cache(maxsize=128)
```
Comment (kinnala) [inline skfem/assembly/basis/basis.py:106]: What is the rationale for `maxsize=128`?

**Comment by kinnala** [comment]
> Where should the HashableNdArray be put? The roof folder is probably not the right location.

This location is fine for now.

**Comment by kinnala** [comment]
I did similar testing of initializing `InteriorBasis` as was done in #586 but the timings here seem to be equal, and assembly is faster, so I think this is good to go.

**Pull request merged**